### ISSUE TITLE
Review Abstract

### ISSUE THREAD

**Issue by smiths** [smiths/spreadOverTimeAndTeam#3]
If you have time, please provide your feedback on the abstract for our paper "A Software Engineering Capstone Infrastructure that Encourages Spreading Work Over Time and Team".

I'll incorporate any feedback or changes I get before 11:00 pm tonight (Oct 2) and submit before 8:00 am tomorrow morning (just in case this is still required.)  We will have an opportunity to change the abstract for the final submission.

***

How can instructors facilitate spreading out the work in a software engineering or computer science capstone course across time and between team members? Currently teams often compromise the quality of their learning experience by frantically working before each deliverable. Some team members further compromise learning by not contributing their fair share to the team effort. To mitigate these problems, we propose using a GitHub template that contains all the infrastructure that is reusable for all teams. All teams will populate their initial repositories with the same folder structure, text-based template documents and template issues. In addition, we propose that at the start of the term each team identify specific quantifiable productivity metrics that they will monitor for each team member, such as the count of meetings attended, issues closed and number of commits. Initial data suggests that these steps have an impact. In 2022/23 we observed 50% of commits happening within 3 days of due dates. After partial introducing the above ideas in 2023/24, this number improved to 37%. Going forward we propose a proper experiment where commit data and interview data is compared between teams that use the proposed interventions and those that do not.

**Comment by CSchank** [comment]
Looks good overall to me! Should we mention fairness metrics for teammate fairness?

**Comment by smiths** [comment]
Good catch @CSchank.  Yes, we should mention the fairness metric.  I'll work it in.

**Comment by smiths** [comment]
I added a discussion of fairness to the abstract in 343631d.  I'm not sure which way of measuring fairness is the best.  I think we should think about this further.  I do like that Schankula's fairness index is 0 when only one person contributes, even if there are two people on the team.  Jain's measure will be much higher.  I believe it will be equal to 0.5 (going by memory).  I think Jain's measure implicitly says that if there are more people sharing the work, then there must be more "work" to do.  If there is a large team and one person does all the work, then his fairness index goes down.  With a small team the fairness index is higher is higher, even though the same person did all the work.  I guess the argument could be if there are 2 hours of work that need to be done on a team of 2, then each person should be expected to do 1 hour of work.  If there are 2 hours of work for a team of 200, then each person should be expected to do 0.01 hours of work.  If one person does all the work in the first case this is unfair, but they have only doubled their original expected work.  If one person does all the work in the second case, then their work is 200 times the original expectation.  Thinking of it this way, I like Jain's metric.  The problem is that our teams are small, so we'll never see a metric that looks that unfair with Jain's index.

I think we should look at Jain's original paper.  It would be nice to have a feel for a what a fairness score of 0.7 means.

@CSchank, how hard would it be to compare 2022/23 to 2023/24 for the fairness metrics?  Do we have any preliminary results we could present?  The absolute value of the fairness metric might be hard to interpret, but if we see the metric go up between years, that is a good sign.  I'm not sure of the right way to compare the years.  An average fairness index for each year seems like not the best measure.

If we can't come up with a story here, we can present our fairness index options and call for future work to explore this further.

**Comment by CSchank** [comment]
Hi Dr. Smith, I do want to look into Jain's metric more. But based on what I know about Jain's fairness metric so far, it has sort of three ``tentpole'' values, where `n` is the number of users sharing a resource:

- $\frac{1}{n}$ - worst case: one person out of $n$ is allocated all the pie. Someone else getting anything would be strictly better than this.
- $1$ - when everyone equally shares the resource
- $\frac{k}{n}$ - when $k$ of the $n$ users have equal pie pieces, but the rest have none

The values in-between $\frac{k}{n}$ and 1 seem less well-defined from what I can tell. For instance, if one of the $n-k$ people with no pie gets a bit, then the situation becomes more fair, but what does that number actually tell us? I'm not sure it's clearly defined.

In the fairness metric I'm proposing, we have the following three types of values:

- 0: when one person is allocated the whole pie
- 1: when everyone has an equal amount of the pie
- Values between 0 and 1: the average proportion of pie that someone has that could be given to someone with less pie, normalized by the number of people and the total amount of pie. We call this type of pie "unfairly held" pie. This number can be represented as $\frac{j}{m}$, where $j$ is the number of "unfairly held" units of the pie, and $m$ is the total number of units of the pie.

Sidenote: I thought about your pie-sharing example, in that example, it's unfair to the people with smaller pieces of pie - the number goes down as the "unfairness" in the pie distribution goes up. At first glance what we want to measure _seems_ to be the opposite: bigger "pieces of pie" seem "good". But they are only 'good" to the person. In this case, the fairness of the situation still goes down as the pie pieces become more unfairly shared, but in this case it's less fair for the people with the bigger pieces of pie. So it's the same idea. Which is the conclusion you also came to, but I wanted to write out my thoughts on it. From now one, instead of pie, we'll talk about the amount of work done by someone:

- 0: when one person does all the work
- 1: when everyone does an equal amount of work
- Values between 0 and 1: the average proportion of "fair" work -- work that someone did that could not have been given to someone who did less, normalized by the number of people and the total amount of work done. For instance, 0.9 means that 90% of the commits were considered to be "fair" and could not have been taken on by someone who did less work.

Let's look at two examples:

Team 1
Person A: 10 commits
Person B: 5 commits
Person C: 5 commits
Jain fairness: 0.8889
"Schankula" fairness: 0.75

In this case, the total is 20 commits. The fairness metric is 0.75, because 5/20 commits could have been given to someone
with a smaller number of commits to make it more fair (Person A could have given 5 to Person B or 5 to Person C). We call 5 the number of "unfair commits" on average (5 could be shifted to Person B, and 5 to Person C, so on average 10/2 = 5). 15 commits were made "fairly", or at least "not unfairly" (in this case: 5 per teammate). The unfairness score is 0.25 (5/20) so the fairness score is 1-0.25=0.75 (15/20).

Now let's look at a different example with the same top performer and total work but different distribution. 

Team 2
Person A: 10 commits
Person B: 8 commits
Person C: 2 commits
Jain fairness: 0.7937
"Schankula" Fairness metric: 0.6

At first this seems maybe more fair, but is it really? Perhaps it may seem so from Person A's perspective, since there is someone else doing more work unlike in Team 1, but overall the low-performer is really underperforming. The work is even less spread out than before. Jain's metric concurs, this is less fair.

My fairness metric is computed as 0.6 here. This is because, of the 20 total commits, 8 was the number of unfair commits, normalized per person. This is calculated as follows: 
- 2 could have been shifted from Person A to Person B, and 8 could have been shifted from Person A to Person C to make it more fair. On average, Person A could have shifted 5 commits to teammates ((2+8)/2 teammates), as with Team 1. 
- However, the fairness is lower than in Team 1 because Person B can also shift 6 commits to Person C. Normalized to the number of teammates, Person B could have shifted 3 commits per teammate. In total, this means that 5+3=8 commits were unfair commits, and thus we get an unfairness score of 8/20=0.4 or a fairness score of 12/20=0.6.

As we saw, this metric tends to follow the same basic trend as Jain's, but the advantage appears that this gives a more "real" meaning to the in-between values. Of course, we lose some information!

Which leads to the other neat thing with this approach: we can decompose this into the "unfairness heatmaps" I showed before (this time with more explanation!). Basically, in this table, we subtract the number of commits of the column Person from the row Person, and divide by the total (e.g. (10-8)/20 = 0.1). This allows us to see at a glance which people contributed the most relative to others. A value of 0.1 means that it a value of 0.1 unfair for Person A to do 10 commits versus Person B's 8, or in other words it had a fairness value of 0.9. Negative numbers indicate an underperformance, that row did less work than column (negative unfairness is just 1+fairness). The neat thing is this table is computed in the same loop that computes the overall unfairness value, so we can reconstruct the overall fairness metric from it, as shown.

|                        |  Person A       | Person B        | Person C       |  Average Positive Unfairness Per Teammate |
| ------------- | ------------- | ------------- |------------- |  --------------|
| Person A        | 0                     |       0.1            |      0.4            |    0.25              |
| Person B        | -0.1                 | 0                    |        0.3          |    0.15               |
| Person C        | -0.4               |        -0.3        |    0                   |         0              |
|                        |                        |                        |    Total Average Unfairness     |   0.4    |
|                        |                        |                        |    Total Average Fairness         |   0.6    |

If this were able to be computed on the fly, it may help instructors/TAs/managers/etc identify people who are underperforming.

Again, I really wonder if someone has already done this, as it seems relatively straightforward. I did arrive at it by myself, but it's still very possible it's been done, so I'm a bit reticent to call it "Schankula fairness"! Still neat nonetheless!

To your point about the team sizes, both metrics would go down if the number of people is increased, except that in the case of the "Schankula fairness" metric, if it happens to be that one person did all the work, it stays 0. Technically though, it's 0 divided by a bigger number :-)

I will try to run the fairness of the year before too. It requires a lot of annoying data cleansing. But it would be good to have for sure! I will try to do so this evening.

**Comment by CSchank** [comment]
After some painstaking work making sure people are the same, here are the results. Some of these may be slightly incorrect due to dirty data, but I did my best.

<img width="758" alt="image" src="https://github.com/user-attachments/assets/2d546399-f704-4ff3-a46a-0163c222e82b">

**Comment by smiths** [comment]
Impressive work @CSchank!  You make some great arguments for your fairness measure.  I do like that you can explain what the numbers mean.

I've been working on a mathematical formulation that we could show in the paper of how to calculate the Schankula index.  What I'm writing feels too complicated, so I thought I would ask you if you have a formulation.  :smile: Did you write functional code to calculate it?  With all the averages of sets of differences, it feels very functional to me.  :smile:  You said you used loops, so now that I think about it, you probably used imperative code.  You are also seem to be thinking in terms of sequences, not sets.  I'm not sure which is a better way to write it.  I'd rather write it functionally, but maybe just writing your code, or pseudo code is the better way to go?  Can you share your code?

These comparisons are reminding me of the Analytic Hierarchy Process (AHP).  There might be something there for us.  I'll have to think about it further.  AHP is all about pairwise comparisons.  It also has a well-understood mathematical underpinning, even though it looks somewhat arbitrary on the surface.

We won't have time before the paper is due, but we should talk to Dr. Janicki.  He is an expert on this kind of thing.  I bet he could tell us if Schankula's index is new.  :smile:

For the purpose of the paper, this is an ideas paper, so I think we can give our idea of an apparently new fairness metric.  :smile:

I like how your metric behaves close to a fairness index of 0.  If the commits are [1e4, 1, 1], I believe your index is 1-0.9999, which is close to zero, but not zero.

Your explanation of the heat map, and its connection to the calculation of your index is great.  It makes sense now.  :smile:

With respect to the data, you did an excellent job pulling something out of the dirty data from last year.  (We can put it as a threat to validity that the github names were ambiguous last year.)  I would really like to compare a snapshot of two years ago to last year, like we can do for the timeline of commits.  For two years ago, can you remove the mechatronics teams from the data?  They are different from the SE teams because someone could do very few commits on a TRON team, but still be a big contributor because they worked on the hardware.  We have no way to measure that.

Would it be meaningful to calculate an average fairness index across all software teams from two years ago and compare that to the average fairness index for all teams last year?  To do it properly, I'm sure there are some stats rules and error bars, but to begin with, I think we can look at the averages to see if they might be meaningful.

**Comment by smiths** [comment]
It is probably too convoluted, but I think I came up with a formal representation of the Schankula index.  If we have a set C of commits, like C = {10, 8, 2}, the index is:

$+\left ( c: \mathbb{N} | c \in C \cdot \frac{+(x: \mathbb{N} | x \in C \cdot (c > x \Rightarrow \frac{c-x}{\Sigma C} | 0 ))}{|C|-1} \right )$

I'm using the notation that I used in SE 2AA4, including the notation for a conditional rule.  I'm sure there is a more standard way to write this.  :smile:

We could write it with the function:

normalizedUnfairCommits(c, C) = $(x: \mathbb{N} | x \in C \cdot (c > x \Rightarrow \frac{c-x}{\Sigma C} | 0 ) )$

Since we are only talking about unfair commits, I've only included the cases where a person has more commits than another.  In your matrix above Chris this would just be the positive numbers, the "unfair" numbers.

Now the index is written as:

$+\left ( c: \mathbb{N} | c \in C \cdot \frac{\text{normalizedUnfairCommits}(c, C)}{|C|-1} \right )$

Here we are using the average of the normalizedUnfairCommits.  

I'm trying to use formulas to match the words that you used to describe what you are doing.

Is there a better way to write this?  I also considered writing it with sequences and capturing your heatmap, but that doesn't feel as motivated by the rationale as the set based approach.

It would be neat if we could prove that this index has to lie between 0 and 1.  I know the normalizing and averaging gets us there, but I know that in an informal sense.  :smile:

**Comment by CSchank** [comment]
I was just about to take a crack at this when I saw your comment. This looks as good or probably better than what I would have come up with! Using the strict greater is very clever, I wasn't sure how to handle not including the person themselves in the count!

One way to maybe make this simpler is we could factor out $\frac{1}{\Sigma C}$ from the `normalizedUnfairCommits` function, and put it in the denominator of the main function, making it $(\left|C\right| - 1) \cdot \Sigma C$. I suppose that would then be named `unfairCommits`, or perhaps the more descriptive but wordy `unnormalizedUnfairCommits` instead!

Then the proof for 0-1 becomes a little more attainable. The worst case is that we have something totally unfair, with one person, say person $x$, doing all the work, all $c_x$ commits. In this case, it's obvious that $\Sigma C = c_x$. The strict > would add up $(c_x - 0)$ a total of $(\left|C\right| - 1)$ times. Which would give an unfairness value of 1, since in that case we have $\frac{(\left|C\right| - 1) \cdot (c_x - 0)}{(\left|C\right| - 1) \cdot \Sigma C} = \frac{(\left|C\right| - 1) \cdot c_x}{(\left|C\right| - 1) \cdot c_x}$. Now, anything other than this is going to decrease the unfairness, since it will increase the numerator but not affect the denominator. The best case is that all of them are equal, in which case we have $\frac{0}{(\left|C\right|-1) \cdot \Sigma C}$. Again, not quite a proof, but there's a sketch of a proof. A concept of a proof, if you will. 😁 Probably will be by contradiction. Maybe I can prove it in Agda! 

In summary: the most unfair commits we can possibly have is just the number of commits times the (number of people minus one). Which we then normalize to a number 1. We can't have more unfair commits than there exist commits because if we did, they'd be included in the sum in the denominator anyway.

Unfortunately, I chose Python (boo) for the data analysis just because it was faster to get started and has easy Matplotlib plotting and lots of resources (and training data!) for things like pulling from GitHub. But it certainly led to many situations where I wasted time debugging things I wouldn't have to in a strictly-typed language!

**Comment by CSchank** [comment]
Sidenote: I was able to reduce the problem of people committing with many names by identifying commits using their GitHub usernames instead of their name submitted with the commits, which helped clean up the data quite a lot. Still, some had multiple accounts, but that was more rare. I have a script that checks for possible close names and asks you, and allows you to input more of them too. 

Good point, I will remove Tron teams from the data. What's the best source of which teams were Tron?

**Comment by smiths** [comment]
The TeamComposition.csv file should have this information.

**Comment by CSchank** [comment]
Okay, probably should have just checked there first 😁

**Comment by CSchank** [comment]
This might be a simpler representation. I suppose we can omit the types of $c$ and $x$ if we know the type of $C$. By the way, I assume this works for real values too. We can also use Dr. Farmer's wisdom to avoid needing to specify that the false values are 0.

$$
+\left(c, x \in C \cdot c > x \Rightarrow  \frac {c-x}{(\left|C\right| - 1) \cdot \Sigma C}\right)
$$

I BELIEVE we can compute this in something like $\mathcal{O}(n \cdot log(n))$ time if we sort the values and then do a single linear pass over those to compute the partial sums which would give us what we want. I may be wrong about that though! Jain's can be computed in linear time, so we're probably a bit worse in that regard. But the extra log(n) term isn't too awful, if I can figure that out. Otherwise the naive implementation is quadratic.

**Comment by smiths** [comment]
Very nice!  We could also move the division outside of the sum.

$$
\frac { +(c, x \in C \cdot c > x \Rightarrow  {c-x})} {(\left|C\right| - 1) \cdot \Sigma C}
$$

I think I like the look of how you wrote it better.  This formula gives us an interesting way to look at the unfairness metric.  It is the sum of all the unfair commits normalized by the product of the total commits and the number of other commits.  We look at all the possible pairs of commits and sum the positive differences.  I like this.  Every time someone commits more than someone else this is unfair.  How unfair is the difference in the number of commits.  We sum up all this unfairness and then normalize it and average it.  Cool!

We don't need to care too much about efficiency.  We won't have to do so many calculations that the calculation stage will be any kind of bottleneck.  A nice bonus if you can do the calculation efficiently.  :smile:

**Comment by CSchank** [comment]
I actually quite like how yours looks. It's nice not having a fraction nested so far inside. And it makes it clear what the normalization is, as you say.

One interesting thing for the (as yet theoretical) full paper would be to compare this metric for commits versus lines of code, or other metrics. Many some linear combination of everything. Even more interesting would be to compare against surveys of how the team felt people were contributing, but that couldn't be done with GitHub data alone and would require ethics approval. Something to think about for next year. But it would be good to see if these quantitative measures of fairness correlate with the qualitative feelings of the teams.

**Comment by smiths** [comment]
I agree with your implicit call for a full paper.  I think there is some content here that we can properly mine and present in a longer format paper.  Some really neat stuff has come up in our conversations.  We might end up disappointed, but there might be something new in Schankula's fairness metric.  (Like I said above, once the dust settles, we should talk to Dr. Janicki to get his thoughts.)

Yes, we will eventually need more data, and we will need data that isn't in GitHub.  I hope we can figure out a way to have meaningful conversations with our students.  The expo event does seem like a candidate time for some meaningful interaction.

**Comment by CSchank** [comment]
Hi Dr. Smith, I see you've inserted the figures into the paper. I'll try to make the text bigger so it's easier to read when it's small. Are there other parts you'd like me to do? Things to write for instance?

**Comment by smiths** [comment]
Thank you for your question @CSchank.  I can think of the following tasks right now (we can create issues for them, if that is helpful):

- fix the two timeline of commits figures, as you suggested.  You are right that the text should be bigger.  Also, please remove the horizontal lines of the grid behind the graph.  It will look nicer without them.  Also, please move the legend so that it doesn't overlap the graph. We should remove the parts of the x-axis that have no data to better use the space.  I'd also like us to do an experiment of seeing what the plots look like if the y-axis uses a log scale.  It probably won't be helpful, but the data has a frustrating large range.  You can also remove the graph titles, since we will have a caption.
- If you have time, you could write the text for the timeline line comparison section.  We want to highlight the difference between 2022-23 and 2023-24 to suggest that  the improvement might be due to the interventions.
- If you have time, you could write the details on your new fairness metric.  I re-wrote the formula to not use the $+(\text{dummy} | R \cdot E)$ notation.  I think more of the audience will be familiar with summation using big sigma.
- Produce new plots for the fairness metrics for 2022-23 and 2023-24.  I would  like the plots to only include SE students.  Also, please sort the teams in ascending order of the fairness index.  We might be able to compare these plots and get some kind of insight.  I'd also like to see the average fairness for both years.

Please let me know if you are writing any section of the paper, so I don't start editing that section and cause a merge conflict.

I am done for today, but I will continue to edit the paper tomorrow.

**Comment by smiths** [comment]
This issue turned into a discussion.  :smile:  We can close it now because the original issue has been dealt with and the discussion reached its logical conclusion.
